### ISSUE TITLE
[Question] Early stopping implementation

### ISSUE THREAD

**Issue by rohan-gt** [catboost/catboost#1387]
Does `early_stopping` work when an eval set is not specified? If not is it possible for the algorithm to automatically sample an eval set from the train set itself?

**Comment by kmedved** [comment]
This would be a useful feature actually. This is how sklearn itself has implemented early stopping, in their new [HistGradientBoostingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html).

I can see some advantages to allowing both - specifically that it's currently not possible to run use something like sklearn's `cross_val_score` on catboost with early stopping, since there's no way to pass an eval set this way. The advantage of using `cross_val_score` is that `cross_val_score` is multiprocessed, so you can search more hyperparameters at once.

**Comment by rohan-gt** [comment]
Optuna actually does this for LightGBM through the [LightGBMTunerCV](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.lightgbm.LightGBMTunerCV.html) method. But this is for automatic tuning. Something similar could be useful for CatBoost

**Comment by rohan-gt** [comment]
Yes it has already been implemented for HistGradientBoostingClassifier (https://github.com/scikit-learn/scikit-learn/issues/14303) although they're using `train_test_split`

**Comment by andrey-khropov** [comment]
> Does early_stopping work when an eval set is not specified?

No

> If not is it possible for the algorithm to automatically sample an eval set from the train set itself?

Automatically - no, but you can do it if you specify an `eval_fraction` parameter (#1500 ).
